File: agent/analysis/engine.py
"""AnalysisEngine — queries the history store and runs all five report types.

The engine is a thin coordinator: it calls store.query_* methods, passes the
results to the pure builders in reports.py, and returns both structured data
and Markdown artifacts.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from agent.history.store import HistoryStore, make_store
from agent.analysis.reports import (
    build_digest,
    build_flaky_report,
    build_spikes_report,
    build_area_health_report,
    build_common_failures_report,
    build_regression_candidates_report,
)


@dataclass
class AnalysisResult:
    flaky: list[dict]
    spikes: list[dict]
    area_health: list[dict]
    common_failures: list[dict]
    regression_candidates: list[dict]
    digest_md: str
    artifacts: dict[str, str]


class AnalysisEngine:
    def __init__(
        self,
        store: Optional[HistoryStore] = None,
        db_url: Optional[str] = None,
        ndjson_path: Optional[Path] = None,
    ) -> None:
        self._store = store or make_store(db_url=db_url, ndjson_path=ndjson_path)

    def run(
        self,
        window: int = 30,
        delta: float = 20.0,
        weeks: int = 4,
        min_suites: int = 2,
        min_flake_rate: float = 10.0,
        min_green: int = 5,
        recent_failures: int = 3,
        suite: Optional[str] = None,
    ) -> AnalysisResult:
        from agent.history.detector import detect_regressions

        flaky = self._store.query_flaky(window=window, suite=suite, min_rate=min_flake_rate)

        # For spikes and area health we need run-level data; query_summary gives
        # per-suite aggregate — for cross-suite we query each known suite and pool.
        # When suite is specified, only that suite. Otherwise all suites from the store.
        suites_to_query = [suite] if suite else self._discover_suites()
        spikes_rows: list[dict] = []
        area_rows: list[dict] = []
        for s in suites_to_query:
            summary = self._store.query_summary(suite=s, runs=window * 2)
            spikes_rows.extend(summary.get("runs", []))

        common_rows = self._query_common_failures(suite=suite)

        regression_candidates = self._detect_regressions(
            suite=suite, min_green=min_green, recent_failures=recent_failures
        )

        digest = build_digest(
            flaky=flaky,
            spikes=spikes_rows,
            area_health=area_rows,
            common_failures=common_rows,
            regression_candidates=regression_candidates,
            window=window,
            delta=delta,
            weeks=weeks,
            min_suites=min_suites,
        )

        artifacts = {
            "flaky.md": build_flaky_report(flaky, window=window, min_rate=min_flake_rate),
            "spikes.md": build_spikes_report(spikes_rows, delta=delta),
            "area-health.md": build_area_health_report(area_rows, weeks=weeks),
            "common-failures.md": build_common_failures_report(common_rows, min_suites=min_suites),
            "regression-candidates.md": build_regression_candidates_report(regression_candidates),
            "digest.md": digest,
        }

        return AnalysisResult(
            flaky=flaky,
            spikes=spikes_rows,
            area_health=area_rows,
            common_failures=common_rows,
            regression_candidates=regression_candidates,
            digest_md=digest,
            artifacts=artifacts,
        )

    def _discover_suites(self) -> list[str]:
        # The local store doesn't have a list-suites method; we query summary
        # with an empty suite and extract from whatever runs exist.
        # For the Supabase store a real query would be better; this is
        # good-enough for local use and the Supabase store can override.
        try:
            from agent.history.local_store import LocalHistoryStore
            if isinstance(self._store, LocalHistoryStore):
                records = self._store._read_all()
                return list({r.get("suite", "") for r in records if r.get("suite")})
        except (ImportError, AttributeError):
            pass
        return []

    def _query_common_failures(self, suite: Optional[str]) -> list[dict]:
        # Query failures from test results; local store has all data inline.
        try:
            from agent.history.local_store import LocalHistoryStore
            if isinstance(self._store, LocalHistoryStore):
                records = self._store._read_all()
                rows = []
                for record in records:
                    if suite and record.get("suite") != suite:
                        continue
                    for t in record.get("tests", []):
                        if t.get("status") in ("failed", "flaky") and t.get("error_text"):
                            rows.append({
                                "test_name": t["test_name"],
                                "suite": record.get("suite", ""),
                                "failure_category": t.get("failure_category", "other"),
                                "error_text": t.get("error_text", ""),
                                "run_count": 1,
                            })
                return rows
        except (ImportError, AttributeError):
            pass
        return []

    def _detect_regressions(
        self, suite: Optional[str], min_green: int, recent_failures: int
    ) -> list[dict]:
        from agent.history.detector import detect_regressions
        try:
            from agent.history.local_store import LocalHistoryStore
            if isinstance(self._store, LocalHistoryStore):
                records = self._store._read_all()
                test_names: set[str] = set()
                for record in records:
                    if suite and record.get("suite") != suite:
                        continue
                    for t in record.get("tests", []):
                        test_names.add(t["test_name"])

                candidates = []
                for name in test_names:
                    timeline = self._store.query_timeline(name)
                    result = detect_regressions(timeline, min_green=min_green, recent_failures=recent_failures)
                    if result["is_regression"]:
                        candidates.append({
                            "test_name": name,
                            "suite": timeline[0].get("suite", "") if timeline else "",
                            "area": None,
                            "green_streak": result["green_streak"],
                            "first_failure_commit": result["first_failure_commit"],
                            "recent_failures": recent_failures,
                        })
                return candidates
        except (ImportError, AttributeError):
            pass
        return []

File: ts/src/analysis/engine.ts
/**
 * AnalysisEngine — queries the history store and runs all report types.
 *
 * Faithful TS port of `agent/analysis/engine.py`. Thin coordinator: calls the
 * store's query methods, passes results to the pure builders in reports.ts, and
 * returns structured data plus Markdown artifacts.
 *
 * Fidelity note: as in Python, `areaHealth` is intentionally never populated by
 * `run()` (the Python engine initialises `area_rows = []` and never appends), so
 * the area-health artifact always renders the empty-data message.
 */

import {
  buildAreaHealthReport,
  buildCommonFailuresReport,
  buildDigest,
  buildFlakyTestsReport,
  buildRegressionCandidatesReport,
  buildFailureSpikesReport,
} from './reports.js';
import type {
  AreaHealthRow,
  CommonFailureRow,
  FlakyRow,
  RegressionRow,
  SpikeRow,
} from './rows.js';
import { detectRegressions } from '../history/detector.js';
import type { AsyncHistoryStore } from '../history/async-store.js';
import type { RunRecord } from '../history/record.js';
import { def } from '../util/coalesce.js';

/**
 * A store that additionally exposes raw records — today, only the local NDJSON
 * backend behind `LocalAsyncAdapter`. See `AsyncHistoryStore.readAll` for why
 * this is an optional capability rather than part of the contract proper.
 */
type ReadableStore = AsyncHistoryStore &
  Required<Pick<AsyncHistoryStore, 'readAll'>>;

function isReadable(store: AsyncHistoryStore): store is ReadableStore {
  return typeof store.readAll === 'function';
}

type RawRecord = RunRecord;
type RawTest = NonNullable<RawRecord['tests']>[number];

function isFailureWithError(t: RawTest): boolean {
  return (
    (t.status === 'failed' || t.status === 'flaky') && Boolean(t.error_text)
  );
}

function toCommonFailureRow(record: RawRecord, t: RawTest): CommonFailureRow {
  return {
    test_name: t.test_name,
    suite: def(record.suite, ''),
    failure_category: def(t.failure_category, 'other'),
    error_text: t.error_text as string,
  };
}

export interface AnalysisResult {
  flaky: FlakyRow[];
  spikes: SpikeRow[];
  areaHealth: AreaHealthRow[];
  commonFailures: CommonFailureRow[];
  regressionCandidates: RegressionRow[];
  digestMd: string;
  artifacts: Record<string, string>;
  /**
   * Sections this run could NOT compute, named in report language (#711).
   *
   * Empty on the local store. Non-empty only when the backend does not
   * implement the optional `readAll()` capability, in which case the sections
   * listed here are UNKNOWN rather than clean — their rows are `[]` for want of
   * data access, not for want of findings, and a caller that renders them as an
   * all-clear is reporting a measurement that never happened.
   */
  degraded: string[];
}

/** Report-language names for the sections that need raw-record access. */
const SECTION_SPIKES = 'failure spikes';
const SECTION_COMMON_FAILURES = 'common failures';
const SECTION_REGRESSIONS = 'regression candidates';

/**
 * Digest inputs, each named for the unit it carries (#670, #673).
 *
 * `windowRuns` counts RUNS (not days), `deltaPp` is a PERCENTAGE-POINT increase
 * (not a percentage), and `minFlakeRatePct` is a 0-100 percentage. The names
 * match the report builders they feed and the `--window-runs` / `--delta-pp` /
 * `--min-rate-pct` flags they are bound to, so the unit survives the whole trip
 * from the command line to the rendered threshold.
 */
export interface RunOptions {
  windowRuns?: number;
  deltaPp?: number;
  weeks?: number;
  minSuites?: number;
  minFlakeRatePct?: number;
  minGreen?: number;
  recentFailures?: number;
  suite?: string | null;
}

export class AnalysisEngine {
  constructor(private readonly store: AsyncHistoryStore) {}

  async run(opts: RunOptions = {}): Promise<AnalysisResult> {
    const windowRuns = opts.windowRuns ?? 30;
    const deltaPp = opts.deltaPp ?? 20.0;
    const weeks = opts.weeks ?? 4;
    const minSuites = opts.minSuites ?? 2;
    const minFlakeRatePct = opts.minFlakeRatePct ?? 10.0;
    const minGreen = opts.minGreen ?? 5;
    const recentFailures = opts.recentFailures ?? 3;
    const suite = opts.suite ?? null;

    const flaky = (await this.store.queryFlaky(
      windowRuns,
      suite,
      minFlakeRatePct,
    )) as FlakyRow[];

    // Sections that need raw-record access are UNKNOWN, not empty, on a backend
    // that does not offer it (#711). `flaky` above and a suite-scoped `spikes`
    // below ride real contract methods, so they stay measured either way.
    const degraded: string[] = [];
    const readable = isReadable(this.store);

    const suitesToQuery = suite ? [suite] : await this.discoverSuites();
    // With no explicit --suite, the suite list itself comes from raw records —
    // so spikes is only degraded in that case, not whenever readAll is absent.
    if (!readable && !suite) degraded.push(SECTION_SPIKES);
    const spikesRows: SpikeRow[] = [];
    for (const s of suitesToQuery) {
      const summary = await this.store.querySummary(s, windowRuns * 2);
      for (const row of def(summary.runs, [])) {
        // query_summary rows omit suite; the spikes builder groups by it, so
        // tag each pooled row with the suite it came from (matches Python).
        spikesRows.push({
          suite: s,
          timestamp: row.timestamp,
          total: row.total,
          failed: row.failed,
          flaky: row.flaky,
        });
      }
    }

    // Faithful to Python: area rows are never populated by run().
    const areaRows: AreaHealthRow[] = [];

    const commonRows = await this.queryCommonFailures(suite);
    if (!readable) degraded.push(SECTION_COMMON_FAILURES);
    const regressionCandidates = await this.detectRegressionCandidates(
      suite,
      minGreen,
      recentFailures,
    );
    if (!readable) degraded.push(SECTION_REGRESSIONS);

    const digest = buildDigest({
      flaky,
      spikes: spikesRows,
      areaHealth: areaRows,
      commonFailures: commonRows,
      regressionCandidates,
      windowRuns,
      deltaPp,
      weeks,
      minSuites,
    });

    const artifacts: Record<string, string> = {
      'flaky.md': buildFlakyTestsReport(flaky, windowRuns, minFlakeRatePct),
      'spikes.md': buildFailureSpikesReport(spikesRows, deltaPp),
      'area-health.md': buildAreaHealthReport(areaRows, weeks),
      'common-failures.md': buildCommonFailuresReport(commonRows, minSuites),
      'regression-candidates.md':
        buildRegressionCandidatesReport(regressionCandidates),
      'digest.md': digest,
    };

    return {
      flaky,
      spikes: spikesRows,
      areaHealth: areaRows,
      commonFailures: commonRows,
      regressionCandidates,
      digestMd: digest,
      artifacts,
      degraded,
    };
  }

  async discoverSuites(): Promise<string[]> {
    if (!isReadable(this.store)) return [];
    const suites = new Set<string>();
    for (const r of await this.store.readAll()) {
      if (r.suite) suites.add(r.suite);
    }
    return [...suites];
  }

  async queryCommonFailures(suite: string | null): Promise<CommonFailureRow[]> {
    if (!isReadable(this.store)) return [];
    const rows: CommonFailureRow[] = [];
    for (const record of await this.store.readAll()) {
      if (suite && record.suite !== suite) continue;
      for (const t of def(record.tests, [])) {
        if (isFailureWithError(t)) rows.push(toCommonFailureRow(record, t));
      }
    }
    return rows;
  }

  async detectRegressionCandidates(
    suite: string | null,
    minGreen: number,
    recentFailures: number,
  ): Promise<RegressionRow[]> {
    if (!isReadable(this.store)) return [];
    const testNames = new Set<string>();
    for (const record of await this.store.readAll()) {
      if (suite && record.suite !== suite) continue;
      for (const t of def(record.tests, [])) testNames.add(t.test_name);
    }

    const candidates: RegressionRow[] = [];
    for (const name of testNames) {
      const timeline = await this.store.queryTimeline(name);
      const result = detectRegressions(timeline, minGreen, recentFailures);
      if (result.is_regression) {
        candidates.push({
          test_name: name,
          suite: timeline.length > 0 ? timeline[0]!.suite : '',
          area: null,
          green_streak: result.green_streak,
          ...(result.first_failure_commit
            ? { first_failure_commit: result.first_failure_commit }
            : {}),
        });
      }
    }
    return candidates;
  }
}
